File: find-the-longest-substring-containing-vowels-in-even-counts/find-the-longest-substring-containing-vowels-in-even-counts.cpp
class Solution {
public:
    //vowel index
map<char, int>vi{
  {'a', 0},
  { 'e',1 },
  { 'i',2 },
  { 'o',3 },
  { 'u',4 }
};
int findTheLongestSubstring(string s) {
  map<int, int>maskIndex;
  int bm = 0;
  int ans = 0;
  maskIndex[bm] = -1;
  for (int i = 0;i < s.size();i++) {
    if (vi.count(s[i])) {
      bm = bm ^ (1 << vi[s[i]]);
    }
    if (!maskIndex.count(bm)) {
      maskIndex[bm] = i;
    }
    ans = max(ans, i - maskIndex[bm]);
  }
  return ans;
}
};

File: 1473-find-the-longest-substring-containing-vowels-in-even-counts/find-the-longest-substring-containing-vowels-in-even-counts.java
class Solution {
    public int findTheLongestSubstring(String s) {
    int n = s.length();
    Map<Character,Integer> vmp = Map.of('a',0,'e',1,'i',2,'o',3,'u',4);
    int bm = 0;
    int ans = 0;
    Map<Integer,Integer> mp = new HashMap<>();
    mp.put(0,-1);
    for(int i = 0;i<n;i++){
      if(vmp.containsKey(s.charAt(i))) bm = bm ^ (1<<vmp.get(s.charAt(i)));
     if(mp.containsKey(bm)){
      ans = Math.max(ans, i - mp.get(bm));
     }else{
       mp.put(bm,i);
     }
    }
    return ans;
    }
}